bqskit/passes/partitioning/quick.py

```diff
@@ -378,3 +425,68 @@ def __init__(
         # Close the bin
         for q in location:
             self.active_qudits.remove(q)
+
+
+class MeasurementBin(Bin):
+    """A special bin made to mark and preserve measurement location."""
+
+    def __init__(
+            self,
+            point: CircuitPoint,
+            location: CircuitLocation,
+            circuit: Circuit,
+    ) -> None:
+        """Initialize a MeasurementBin with the point and location of a
+        measurement."""
+        super().__init__()
+
+        # Add the measurement
+        self.add_op(point, location)
+
+        # Measurement bins fill the volume to the next gates
+
+        nexts = circuit.next(point)
+        ends: dict[int, int | None] = {q: None for q in location}
+        for p in nexts:
+            loc = circuit[p].location
+            for q in loc:
+                if q in ends and (ends[q] is None or ends[q] >= p.cycle):  # type: ignore # noqa # short-circuit safety for >=
+                    ends[q] = p.cycle - 1
+        self.ends = ends
+
+        # Close the bin
+        for q in location:
+            self.active_qudits.remove(q)
+
+
+class ResetBin(Bin):
+    """A special bin made to mark and preserve reset location."""
+
+    def __init__(
+            self,
+            point: CircuitPoint,
+            location: CircuitLocation,
+            circuit: Circuit,
+    ) -> None:
+        """Initialize a reset with the point and location of a reset."""
+        super().__init__()
+
+        # Add the reset
+        self.add_op(point, location)
+
+        # Reset bins fill the volume to the next gates
+
+        nexts = circuit.next(point)
+        ends: dict[int, int | None] = {q: None for q in location}
+        for p in nexts:
+            loc = circuit[p].location
+            for q in loc:
+                if q in ends and (
+                        ends[q] is None or ends[q] >= p.cycle  # type: ignore # noqa # short-circuit safety for >=
+                ):
+                    ends[q] = p.cycle - 1
+        self.ends = ends
+
+        # Close the bin
+        for q in location:
+            self.active_qudits.remove(q)
```

Comment:
Same as above... Do we really need to define 3 classes with the exact same functionality, code, and usage? Is there a way to abstract this and reduce on repetitive code?